### PR TITLE
Build in container and update CI

### LINKED ISSUE #1392: Suggestion: OCI predefined image labels
The Open Container Initiative (OCI) has a number of predefined image labels that are commonly used in automation to better understand the published image, outlined [here](https://github.com/opencontainers/image-spec/blob/main/annotations.md).

Adding these annotations to the images before they're published helps users (especially those who use automation) better consume these images. Its a small change to either the dockerfile (to add LABEL instruction) or to the build step.

I am happy to work on this as it's relatively simple.

### PR DESCRIPTION
**Pull request by jtcarnes**
This MR changes the build process for metallb to build inside a docker container instead of building outside and copying in. It further refactors much of the CI to better build images inside containers and leverage github actions instead of custom scripting.

Fixes #1392 #1408 #1413

### CODE REVIEW

File: .github/workflows/release-chart.yaml [hunk 1]
```diff
@@ -10,7 +10,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - name: Checkout
-        uses: actions/checkout@v2
+        uses: actions/checkout@v3
```
Comment (fedepaol) [inline .github/workflows/release-chart.yaml:13]: This doesn't feel it belongs to this commit. Maybe a general (separate) commit to bump up the version of this action?

File: .github/workflows/publish.yaml [hunk 2]
```diff
@@ -48,39 +47,47 @@ jobs:
...
+            type=semver,pattern={{version}}
+          labels: |
+            org.opencontainers.image.title=${{ matrix.image }}
+            org.opencontainers.image.description=${{ matrix.image }} for metallb, a network load-balancer implementation for Kubernetes using standard routing protocols
+
+      - name: Build and push ${{ matrix.image }}
+        uses: docker/build-push-action@v3
+        with:
+          context: .
+          tags: ${{ steps.meta.outputs.tags }}
```
Comment (fedepaol) [inline .github/workflows/publish.yaml:88]: Should we inject GIT_BRANCH and GIT_COMMIT when we build the images?

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -56,31 +55,42 @@ jobs:
...
-          inv build --binaries all
-          docker save quay.io/metallb/speaker:dev-amd64 > speaker.tar
-          docker save quay.io/metallb/controller:dev-amd64 > controller.tar
-          docker save quay.io/metallb/mirror-server:dev-amd64 > mirror-server.tar
+  build-test-images:
+    runs-on: ubuntu-20.04
+    strategy:
+      fail-fast: true
+      matrix:
+        image: [speaker, controller, mirror-server]
```
Comment (fedepaol) [inline .github/workflows/ci.yaml:63]: As before, I'd remove mirror-server (and add the newly added [configmaptocrs](https://github.com/metallb/metallb/tree/main/configmaptocrs) image

**Comment by fedepaol** [comment]
Neat! Thanks a lot!

File: mirror-server/Dockerfile [hunk 4]
```diff
@@ -1,4 +1,15 @@
+# syntax=docker/dockerfile:1
```
[the remark this answers is not in the document]
Comment (jtcarnes) [reply mirror-server/Dockerfile:1]: I'll delete it if you want, but currently its on main. What do you think?

File: controller/Dockerfile [hunk 5]
```diff
@@ -13,7 +13,7 @@ COPY controller/*.go .
 COPY internal internal
 COPY api api
 RUN CGO_ENABLED=0 GOARM=6 go build -v -o /build/controller \
-  -ldflags "'-X go.universe.tf/metallb/internal/version.gitCommit=${GIT_COMMIT} -X go.universe.tf/metallb/internal/version.gitBranch=${GIT_BRANCH} -X'"
+  -ldflags "-X go.universe.tf/metallb/internal/version.gitCommit=${GIT_COMMIT} -X go.universe.tf/metallb/internal/version.gitBranch=${GIT_BRANCH} -X"
```
[the remark this answers is not in the document]
Comment (jtcarnes) [reply controller/Dockerfile]: I kept each dockerfile change on a different commit, but I can squash all the dockerfile changes into one if that's preferred.

File: controller/Dockerfile [hunk 6]
```diff
@@ -12,8 +12,10 @@ RUN go mod download
 COPY controller/*.go .
 COPY internal internal
 COPY api api
-RUN CGO_ENABLED=0 GOARM=6 go build -v -o /build/controller \
-  -ldflags "-X go.universe.tf/metallb/internal/version.gitCommit=${GIT_COMMIT} -X go.universe.tf/metallb/internal/version.gitBranch=${GIT_BRANCH} -X"
+
+# Cache builds directory for faster rebuild
+RUN --mount=type=cache,target=/root/.cache/go-build CGO_ENABLED=0 GOARM=6 go build -v -o /build/controller \
```
[the remark this answers is not in the document]
Comment (jtcarnes) [reply controller/Dockerfile:17]: See above

File: .github/workflows/publish.yaml (hunk 2 above)
Comment (jtcarnes) [reply .github/workflows/publish.yaml:88]: Good idea, fixed

File: .github/workflows/release-chart.yaml (hunk 1 above)
Comment (jtcarnes) [reply .github/workflows/release-chart.yaml]: I've removed it for now. Can do in a different MR.

File: .github/workflows/ci.yaml (hunk 3 above)
Comment (jtcarnes) [reply .github/workflows/ci.yaml:63]: I'll add a commit for configmaptocrs image. As for mirror-server, do you want to remove it entirely, or something else?

Comment (fedepaol) [reply .github/workflows/ci.yaml]: See above

File: mirror-server/Dockerfile (hunk 4 above)
[the remark this answers is not in the document]
Comment (jtcarnes) [reply mirror-server/Dockerfile:1]: So, leave it for now or remove it?

[the remark this answers is not in the document]
Comment (fedepaol) [reply mirror-server/Dockerfile:1]: leave it for now

File: .github/workflows/composite/setup/action.yaml [hunk 7]
```diff
@@ -25,13 +26,13 @@ runs:
...
       shell: bash
       working-directory: metallb-images
       run: |
-        docker load -i speaker.tar
-        docker load -i controller.tar
-        docker load -i mirror-server.tar
+        docker load -i image-tar-speaker/speaker.tar
+        docker load -i image-tar-controller/controller.tar
+        docker load -i image-tar-mirror-server/mirror-server.tar
+        docker load -i image-tar-configmaptocrs/configmaptocrs.tar
```
Comment (oribon) [inline .github/workflows/composite/setup/action.yaml:38]: I don't think we use this anywhere, we can skip this? @fedepaol

File: .github/workflows/publish.yaml [hunk 8]
```diff
@@ -48,39 +47,51 @@ jobs:
...
           inv checkchanges --action="run inv generateapidocs"
 
-  build-and-push:
-    needs: unit-tests
+  publish-images:
     runs-on: ubuntu-20.04
+    strategy:
+      fail-fast: true
+      matrix:
+        image: [speaker, controller, mirror-server, configmaptocrs]
```
Comment (oribon) [inline .github/workflows/publish.yaml:55]: We haven't published mirror-server until now, and as @fedepaol said we might be able to remove it completely (in a different pr ofc)

**Review by oribon — COMMENTED** [review]
this looks great! thanks a lot for filing

File: .github/workflows/composite/setup/action.yaml (hunk 7 above)
Comment (fedepaol) [reply .github/workflows/composite/setup/action.yaml:38]: I think we do

Comment (jtcarnes) [reply .github/workflows/composite/setup/action.yaml:38]: They are used in the e2e testing, if i´m not mistaken. You load the images so when you do you´re deploying, you´re loading the pipelines image

File: .github/workflows/publish.yaml (hunk 8 above)
Comment (jtcarnes) [reply .github/workflows/publish.yaml:55]: Ok, can remove

**Comment by fedepaol** [comment]
Mind checking the backward compatible lane? I see it's constantly failing. Thanks!

**Comment by jtcarnes** [comment]
@fedepaol looks like the check error should be fixed now. Give it a once over again when you get the chance, if you get the time.

**Comment by jtcarnes** [comment]
@fedepaol  I believe theres currently an expected check on unit-tests-and-build, which I have removed in favor of doing image build and testing in different steps (build-test-images (IMAGE) and unit-test). Because at the repo settings level its expecting a report from unit-tests-and-build, it will wait forever (job doesn't exist). I cannot change that in the pipeline as it's set by the repo instead of by the CI. If you accept the MR, make sure you change the expected check setting on the repo.

Discussion example can be found [here](https://github.community/t/expected-waiting-for-status-to-be-reported/18001)

File: .github/workflows/composite/setup/action.yaml (hunk 7 above)
Comment (fedepaol) [reply .github/workflows/composite/setup/action.yaml:38]: Not the `mirror-server` image I think

Comment (jtcarnes) [reply .github/workflows/composite/setup/action.yaml:38]: If mirror-server is not used in testing and not published, where is it used? These loads are only for testing.

Comment (fedepaol) [reply .github/workflows/composite/setup/action.yaml:38]: It's what we discussed here https://github.com/metallb/metallb/pull/1414#discussion_r892413347 but I wasn't totally sure so we agreed to keep it for now.
I don't think it is used at all, so if you want you can remove everything mirror-server related in this PR (in a separate commit) or we can handle it later

Comment (jtcarnes) [reply .github/workflows/composite/setup/action.yaml:38]: I would say for now, to keep things atomic and easier to comprehend, lets just leave it and remove in another pr.
(ie this mr changes everything to a different method, another mr removes unneeded components)

Comment (fedepaol) [reply .github/workflows/composite/setup/action.yaml:38]: Agree

Comment (jtcarnes) [reply .github/workflows/composite/setup/action.yaml:38]: I have a separate MR ready for it, #1436 , so I dont forget.

**Comment by fedepaol** [comment]
@jtcarnes mind rebasing this?

**Comment by fedepaol** [comment]
> @fedepaol ok, should be gtg now

great, let's wait for CI

**Comment by fedepaol** [comment]
LGTM, thanks!

**Comment by fedepaol** [comment]
@jtcarnes looks like the publish jobs are failing (see https://github.com/metallb/metallb/runs/6933490643?check_suite_focus=true ) for example. Can you have a look?

**Pull request merged**